File: 코드/ch5/clip7/boj16987/main.py
import sys

class Egg:
    def __init__(self, durability, weight):
        self.durability = durability
        self.weight = weight

    def fight(self, other):
        self.durability -= other.weight
        other.durability -= self.weight

    def restore(self, other):
        self.durability += other.weight
        other.durability += self.weight

def solve(pick):
    global answer
    if pick == n:
        count = sum(1 for egg in eggs if egg.durability <= 0)
        answer = max(answer, count)
        return
    
    if eggs[pick].durability > 0:
        target_exists = False
        for target in range(n):
            if target == pick:
                continue
            if eggs[target].durability > 0:
                target_exists = True
                eggs[pick].fight(eggs[target])
                solve(pick + 1)
                eggs[pick].restore(eggs[target])
        if not target_exists:
            solve(pick + 1)
    else:
        solve(pick + 1)

n = int(sys.stdin.readline())
eggs = []
for _ in range(n):
    durability, weight = map(int, sys.stdin.readline().split())
    eggs.append(Egg(durability, weight))

answer = 0
solve(0)
print(answer)

File: 코드/ch5/clip7/boj16987/Main.java
import java.util.Scanner;

public class Main {


    public static int n;
    public static Egg[] eggs;
    public static int answer = 0;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        eggs = new Egg[n];
        for(int i = 0; i < n; i++) {
            eggs[i] = new Egg(sc.nextInt(), sc.nextInt());
        }
        solve(0);
        System.out.println(answer);
    }
    public static void solve(int pick) {
        if(pick == n) {
            int count = 0;
            for(int i = 0; i < n; i++) {
                if(eggs[i].durability <= 0) count++;
            }
            answer = Math.max(answer, count);
            return;
        }

        if(eggs[pick].durability > 0) {
            boolean targetExists = false;
            for(int target = 0; target < n; target++) {
                if(target == pick) continue;
                if(eggs[target].durability > 0) {
                    targetExists = true;
                    eggs[pick].fight(eggs[target]);
                    solve(pick + 1);
                    eggs[pick].restore(eggs[target]);
                }
            }
            if(!targetExists) solve(pick + 1);
        }
        else {
            solve(pick + 1);
        }
    }
}

class Egg {
    int durability;
    int weight;

    public Egg(int durability, int weight) {
        this.durability = durability;
        this.weight = weight;
    }

    public void fight(Egg other) {
        this.durability -= other.weight;
        other.durability -= this.weight;
    }

    public void restore(Egg other) {
        this.durability += other.weight;
        other.durability += this.weight;
    }
}
